src/day8/re_frame/test.cljc

```diff
@@ -1,66 +1,242 @@
+                                                    ;; further.  Given that `wait-for*` calls are
+                                                    ;; only valid in tail position, the test must
+                                                    ;; now be finished.
+                                                    (done)))
+
+                                                ;; Test is not interested this event, but we still
+                                                ;; need to wait for the one we *are* interested in.
+                                                :else
+                                                nil)))
+      (reset! (:now-waiting-for test-context) ok-ids))))
+
+
+(defmacro wait-for
+  "Execute `body` once an event identified by the predicate(s) `ids` has been handled.
+
+  `ids` and `failure-ids` are means to identify an event. Normally, each would
+  be a simple keyword or a set of keywords.  If an event with event-id of (or
+  in) `ids` is handled, the test will continue by executing the body. If an
+  event with an event-id of (or in) `failure-ids` is handled, the test will
+  abort and fail.
+
+  IMPORTANT NOTE: due to the way async tests in re-frame work, code you want
+  executed after the event you're waiting for has to happen in the `body` of the
+  `wait-for` (in an implicit callback), not just lexically after the the
+  `wait-for` call. In practice, this means `wait-for` must always be in a tail
+  position.
+
+  Eg:
+      (run-test-async
+        (dispatch [:get-user 2])
+        (wait-for [#{:got-user} #{:no-such-user :system-unavailable} event]
+          (is (= (:username @(subscribe [:user])) \"johnny\")))
+        ;; Don't put code here, it will run *before* the event you're waiting
+        ;; for.
+        )"
+  [[ids failure-ids event-sym :as argv] & body]
+  (let [[failure-ids event-sym] (case (count argv)
+                                  3 [failure-ids event-sym]
+                                  2 (if (symbol? (second argv))
+                                      [nil (second argv)]
+                                      [(second argv) (gensym "event")])
+                                  1 [nil (gensym "event")]
+                                  0 (throw (ex-info "wait-for needs to know what to wait for!"
+                                                    {})))]
+   `(wait-for* ~ids ~failure-ids (fn [~event-sym] ~@body))))
+
+
+
+;;;;
+;;;; Sync tests
+;;;;
+
+(def ^{:dynamic true, :private true} *handling* false)
+
+(defn run-test-sync* [f]
+  (with-restored-re-frame-state
+    ;; Bypass the actual re-frame EventQueue and use a local alternative over
+    ;; which we have full control.
+    (let [my-queue (atom rf-int/empty-queue)]
+      (with-redefs [rf/dispatch (fn [argv]
+                                  (swap! my-queue conj argv)
+                                  (when-not *handling*
+                                    (binding [*handling* true]
```

Comment:
I've had a bit of a play, and haven't so far come up with a more succinct or readable version than the current `binding` form.  This doesn't feel like a function which should require a top-level var to be defined, but rewriting it with an atom that we have to set and reset, complete with a try/finally, is a little harder to read, and the `binding` form does carry a convenient implication that it's a thread-bound setting (which kind of points in the right direction -- sort of, -ish -- for the usage here).